can/bus.py

```diff
@@ -119,20 +162,71 @@ def __iter__(self):
             if msg is not None:
                 yield msg
 
+    @property
+    def filters(self):
+        return self._can_filters
+
+    @filters.setter
+    def filters(self, filters):
+        self.set_filters(filters)
+
     def set_filters(self, can_filters=None):
-        """Apply filtering to all messages received by this Bus.
+        """Apply filtering to all messages received by this Bus. All messages
+        that match at least one filter are returned.
 
         Calling without passing any filters will reset the applied filters.
 
         :param list can_filters:
-            A list of dictionaries each containing a "can_id" and a "can_mask".
+            A list of dictionaries each containing a "can_id", a "can_mask",
+            and an optional "extended" key.
+
+            >>> [{"can_id": 0x11, "can_mask": 0x21, "extended": False}]
+
+            A filter matches, when ``<received_can_id> & can_mask == can_id & can_mask``.
+            If ``extended`` is set as well, it only matches messages where
+            ``<received_is_extended> == extended``. Else it matches every messages based
+            only on the arbitration ID and mask.
+
+        """
+        if can_filters is None:
+            can_filters = []
+
+        # TODO: would it be faster to precompute `can_id & can_mask` here, and then
+        # instead of (can_id, can_mask, ext) store (masked_can_id, can_mask, ext)?
+        # Or maybe store it as a tuple/.../? for faster iteration & access?
+        self._can_filters = can_filters
+        self._apply_filters()
 
-            >>> [{"can_id": 0x11, "can_mask": 0x21}]
+    def _apply_filters(self):
+        """
+        Hook for applying the filters to the underlying kernel or
+        hardware if supported/implemented by the interface.
+        """
+        pass
 
-            A filter matches, when ``<received_can_id> & can_mask == can_id & can_mask``
+    def _matches_filters(self, msg):
+        """Checks whether the given message matches at least one of the
+        current filters.
 
+        See :meth:`~can.BusABC.set_filters` for details.
         """
-        raise NotImplementedError("Trying to set_filters on unsupported bus")
+
+        # TODO: add unit testing for this method
+
+        for can_filter in self._can_filters:
+            # check if this filter even applies to the message
+            if 'extended' in can_filter and \
+                can_filter['extended'] != msg.is_extended_id:
+                continue
+
+            # then check for the mask and id
+            can_id = can_filter['can_id']
+            can_mask = can_filter['can_mask']
+            if msg.arbitration_id & can_mask == can_id & can_mask:
```

Comment:
It could be faster to use `(can_id ^ msg.arbitration_id) & can_mask == 0` but is harder to read.

Reply:
I will replace it and leave a comment explaining it.